### DISCUSSION TITLE
How to specify which order should run the build task of package

### DISCUSSION THREAD

**Discussion by tgmarinho** [vercel/turborepo#1347] [Help] [answered]
how to set up the turborepo to run the build scripts from each package.json of the packages repo in some specific order?

I have the packages: a, b, c, I wanna run the build in this order: b,c, a

I think the answer is here: https://turborepo.org/docs/core-concepts/pipelines#implicit-dependencies-and-specific-package-tasks But I couldn't figure out it yet, my suggestion, if possible, is to have some example package with this setup.

https://twitter.com/tgmarinho/status/1534272696963538946

**Comment by ThibautMarechal** [comment, marked as answer]
You have 2 options: 

## 1) Make that link visible in each package dependencies

_(C depends on B and A depdends on C)_  
```
{
  "name": "a",
  "dependencies": {
    "c": "*"
  },
  "scripts": {
    "build": "echo a"
  }
}
```
```
{
  "name": "b",
  "scripts": {
    "build": "echo b"
  }
}
```
```
{
  "name": "c",
  "dependencies": {
    "b": "*"
  },
  "scripts": {
    "build": "echo c"
  }
}
```

_Then you just need to configure the pipelines with the ```^``` before the task's name._
```
{
  "$schema": "https://turborepo.org/schema.json",
  "pipeline": {
    "build": {
      "dependsOn": ["^build"]
    }
  }
}
```

## 2) Make that link visible in the ```turbo.json``` file

_(no package depends on each others)_  
```
{
  "name": "a",
  "scripts": {
    "build": "echo a"
  }
}
```
```
{
  "name": "b",
  "scripts": {
    "build": "echo b"
  }
}
```
```
{
  "name": "c",
  "scripts": {
    "build": "echo c"
  }
}
```

_(build of C depends on build of B and build of A depdends on build of C)_  
```
{
  "$schema": "https://turborepo.org/schema.json",
  "pipeline": {
    "a#build": {
      "dependsOn": ["c#build"]
    },
    "b#build": {},
    "c#build": {
      "dependsOn": ["b#build"]
    }
  }
}
```

I personally prefer the first solution, but I guess it depends on the use case 😄

**Reply by Neo-Ciber94** [reply]
The 1st solution didn't work for me, but the second one does.

Thank!

**Comment by tgmarinho** [comment]
https://www.tgmarinho.com/posts/2022-06-09-deploy-monorepo-turborepo-heroku

**Comment by arkmech** [comment]
I tried this out with dev script `tsup --watch`, but it stalls on the dependent package because I guess turborepo can't tell if its done. Since its in watch mode, Even though tsup build outputs: `DTS ⚡️ Build success in 3039ms`

**Reply by 58bits** [reply]
Hi @arkmech  - I'm having the same problem. Did you find a solution? I'd like the package with `tsup --watch` to execute first, and then watch, and then build the other packages.

**Reply by arkmech** [reply]
@58bits, I ended up solving the issue with a workaround. I had the `clean` option in tsup config which only cleans declaration files it seems. It does take a good amount of time to build type declaration files for my other packages that depend on it. This causes all my watch builds to fail.

So I just removed the `clean` option and I'm all set with this issue.

**Reply by 58bits** [reply]
Ah I see - thanks - I'll give it a try.

**Reply by anthonyshew** [reply]
`turbo watch` is unable to know when a built-in task watcher (like `tsup --watch`) has completed it's work. Since it's a persistent task, it has no "end" state.

Instead, you likely want to use `tsup` (without `--watch`) and make sure the task isn't marked as persistent.

**Comment by buster95** [comment]
In the last version of turbo repo pipeline isnt recognized so you can do this

```json
{
    "$schema": "https://turborepo.com/schema.json",
    "ui": "tui",
    "tasks": {
        "build": {
            "outputs": [
                "dist/**"
            ]
        },
        "fronty#build": {
            "dependsOn": [
                "backy#build"
            ]
        },
        "check-types": {
            "dependsOn": [
                "^check-types"
            ]
        },
        "dev": {
            "persistent": true,
            "cache": false
        }
    }
}
```
